src/gen-ts.ts

```diff
@@ -8,295 +8,184 @@
-    element: Element, declarations: string[], indent: number = 0) {
-  if (!element.className) {
-    // TODO: handle elements with tagnae, but no exported class
-    return;
+/**
+ * Extend the given TypeScript declarations document with all of the relevant
+ * items in the given Polymer Analyzer document.
+ */
+function handleDocument(doc: analyzer.Document, root: ts.Document) {
+  // TODO Should we traverse and serialize all features in the same order they
+  // were originally declared, instead of grouping by type as we do here?
+  for (const element of doc.getFeatures({kind: 'element'})) {
+    handleElementOrBehavior(element, root);
   }
-  const {name: className, namespace: namespaceName} =
-      getNamespaceAndName(element.className);
-  if (namespaceName !== 'Polymer') {
-    // TODO: handle non-Polymer namespaces
-    return;
+  for (const behavior of doc.getFeatures({kind: 'behavior'})) {
+    handleElementOrBehavior(behavior, root);
   }
+}
 
-  const node = element.astNode;
-  let d = '';
-
-  if (node == null) {
-    process.stderr.write(`no AST node for ${element.className}\n`);
-    d += `${idt(indent)}class ${className} {\n`;
-  } else if (
-      node.type === 'ClassDeclaration' || node.type === 'ClassExpression') {
-    d += `${idt(indent)}class ${node.id!.name} extends ${
-        generate(node.superClass)} {\n`;
-  } else if (node.type === 'VariableDeclaration') {
-    // We can't output the variable initializer, so we union the mixins
-    let supers: string[] = [];
-    if (element.superClass) {
-      const superName =
-          getNamespaceAndName(element.superClass.identifier).name!;
-      supers.push(superName)
-    }
-    for (const mixin of element.mixins) {
-      const mixinName = getNamespaceAndName(mixin.identifier).name!;
-      supers.push(mixinName)
-    }
-    const type = supers.map((s) => `Constructor<${s}>`).join('&');
-
-    d += `${idt(indent)}const ${className}: ${type};\n`;
-    declarations.push(d);
-    return;
+/**
+ * Add the given Element or Behavior to the given TypeScript declarations
+ * document.
+ */
+function handleElementOrBehavior(
+    feature: analyzer.Element|analyzer.PolymerBehavior, root: ts.Document) {
+  let className, parent;
+  if (feature.className) {
+    const parts = feature.className.split('.');
+    className = parts[parts.length - 1];
+    parent = findOrCreateNamespace(root, parts.slice(0, -1));
+  } else if (feature.tagName) {
+    className = kebabToCamel(feature.tagName);
+    parent = root;
   } else {
-    process.stderr.write(`no AST node for ${element.className}\n`);
-    d += `${idt(indent)}class ${className} {\n`;
+    console.error('Could not find a name.');
+    return;
   }
 
-  for (const [, property] of element.properties) {
-    if (property.privacy === 'private' || property.inheritedFrom != null) {
-      continue;
+  const extendz = new Array<string>();
+  if (isPolymerElement(feature)) {
+    extendz.push('Polymer.Element');
```

Comment:
TODO: only do this when it's a legacy polymer element?

Reply:
Added TODO.